File: core/src/main/java/by/mrrockka/features/calculation/CalculationService.java
package by.mrrockka.features.calculation;

import by.mrrockka.domain.game.Game;
import by.mrrockka.domain.payout.Payout;
import by.mrrockka.service.GameService;
import by.mrrockka.service.MoneyTransferService;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Isolation;
import org.springframework.transaction.annotation.Propagation;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;


@Service
@RequiredArgsConstructor
public class CalculationService {

  private final List<CalculationStrategy> strategies;
  private final MoneyTransferService moneyTransferService;
  private final GameService gameService;

  @Transactional(isolation = Isolation.READ_COMMITTED, propagation = Propagation.REQUIRED)
  public List<Payout> calculateAndSave(final Game game) {
    final var payouts = strategies.stream()
      .filter(strategy -> strategy.isApplicable(game))
      .findFirst()
      .orElseThrow(() -> new NoStrategyFoundToCalculateGameException(game))
      .calculate(game);

    if (gameService.doesGameHasUpdates(game)) {
      //  todo: add transactionality service to execute transaction only part of code
      gameService.finishGame(game);
      moneyTransferService.storeBatch(game, payouts);
    }
    return payouts;
  }

}

File: core/src/main/java/by/mrrockka/service/calculation/CalculationService.kt
package by.mrrockka.service.calculation

import by.mrrockka.domain.Game
import by.mrrockka.domain.Payout
import by.mrrockka.service.GameServiceOld
import by.mrrockka.service.MoneyTransferService
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Propagation
import org.springframework.transaction.annotation.Transactional

@Service
open class CalculationService(
        val calculator: GameCalculator,
        val gameService: GameServiceOld,
        val moneyTransferService: MoneyTransferService,
) {
    @Transactional(propagation = Propagation.REQUIRED)
    open fun calculateAndSave(game: Game): List<Payout> {
        val payouts = calculator.calculate(game)
        if (game.finishedAt == null) {
            gameService.finishGame(game)
            moneyTransferService.storeBatch(game, payouts)
        }
        return payouts
    }
}
